Userland/Libraries/LibVirtGPU/Device.cpp

```diff
@@ -1,21 +1,155 @@
 /*
  * Copyright (c) 2022, Stephan Unverwerth <s.unverwerth@serenityos.org>
+ * Copyright (c) 2022, Sahan Fernando <sahan.h.fernando@gmail.com>
+ * Copyright (c) 2022, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #include <AK/NonnullOwnPtr.h>
+#include <Kernel/API/VirGL.h>
 #include <LibCore/File.h>
 #include <LibCore/System.h>
+#include <LibVirtGPU/CommandBufferBuilder.h>
 #include <LibVirtGPU/Device.h>
 #include <LibVirtGPU/Image.h>
 #include <LibVirtGPU/Shader.h>
+#include <LibVirtGPU/VirGLProtocol.h>
 
 namespace VirtGPU {
 
-Device::Device(NonnullRefPtr<Core::File> gpu_file, Gfx::IntSize)
+static constexpr auto frag_shader = "FRAG\n"
+                                    "PROPERTY FS_COLOR0_WRITES_ALL_CBUFS 1\n"
+                                    "DCL IN[0], COLOR, COLOR\n"
+                                    "DCL OUT[0], COLOR\n"
+                                    "  0: MOV OUT[0], IN[0]\n"
+                                    "  1: END\n"sv;
+
+static constexpr auto vert_shader = "VERT\n"
+                                    "DCL IN[0]\n"
+                                    "DCL IN[1]\n"
+                                    "DCL OUT[0], POSITION\n"
+                                    "DCL OUT[1], COLOR\n"
+                                    "DCL CONST[0..3]\n"
+                                    "DCL TEMP[0..1]\n"
+                                    "  0: MUL TEMP[0], IN[0].xxxx, CONST[0]\n"
+                                    "  1: MAD TEMP[1], IN[0].yyyy, CONST[1], TEMP[0]\n"
+                                    "  2: MAD TEMP[0], IN[0].zzzz, CONST[2], TEMP[1]\n"
+                                    "  3: MAD OUT[0], IN[0].wwww, CONST[3], TEMP[0]\n"
+                                    "  4: MOV_SAT OUT[1], IN[1]\n"
+                                    "  5: END\n"sv;
+
+Device::Device(NonnullRefPtr<Core::File> gpu_file, Gfx::IntSize size)
     : m_gpu_file { gpu_file }
 {
+    // Create a virgl context for this file descriptor
+    MUST(Core::System::ioctl(m_gpu_file->fd(), VIRGL_IOCTL_CREATE_CONTEXT));
```

Comment:
One more thing, I don't see us destroying this VirGL context in the destructor of the class (there is no destructor?). Not doing this will make it much harder to unit test.

Reply:
Unfortunately there isn't yet a way to destroy contexts in the kernel driver. I added a FIXME above the `initialize_context()` method.